### DISCUSSION TITLE
How can I simulate hidden service (onion service) using Shadow?

### DISCUSSION THREAD

**Discussion by Allgot** [shadow/shadow#3253] [Q&A] [answered]
Hi!  First of all, we have been studying this wonderful simulator for a while and would like to thank for your hard work in maintaining this project.

My team is working to evaluate the effectiveness of some attacks against hidden services, using Shadow. We could find some hints from this [old archive](https://github.com/shadow/shadow-plugin-tor/tree/main/resource), but it is out-dated and not compatible with the current version of Shadow.

It seems that we can build our configuration file from the ground up by referring to the old archive. However, before doing so, we want to make sure that this is not reinventing the wheel. Is there any updated, publicly available configuration file (i.e., shadow.yaml file)?

Thank you :)

**Comment by Allgot** [comment]
By the way, we have checked several artifacts from recent studies (e.g., [explainwf](https://explainwf-popets2023.github.io/shadow/)), but we couldn't figure out how to configure the hidden service, unfortunately :(

**Comment by robgjansen** [comment, marked as answer]
Yes, you should use [tornettools](https://github.com/shadow/tornettools) to help you generate new Tor network configurations that can run in shadow. There are several steps involved as explained in the README of that project. You can use the `onion` options during the generate step to configure hidden services: see `tornettools generate --help` for more details.

**Reply by Allgot** [reply]
Thank you for introducing these tools for us :)
It seems that we can receive enough aid by using these tools.

Thank you!

**Reply by Allgot** [reply]
Hello @robgjansen, we made every effort to configure our setup correctly, but unfortunately, we encountered difficulties in conducting stable experiments. The issue we are facing is that clients consistently fail to connect to the hidden service when the PoW-based defense mechanism of Tor is active. Specifically, while clients can normally establish and maintain connections to the hidden service when this mechanism is inactive, they fail to maintain their connection after the successful initial connection when the mechanism is activated.

To provide you with a better understanding of our situation, we have attached two images illustrating how the number of connected clients to the hidden service changes over time. These experiments were conducted following the instructions outlined in the readme file available at [this link](https://github.com/shadow/tornettools).

![image](https://github.com/shadow/shadow/assets/96973603/7d3aea80-35ec-4a9c-923b-459f33474b05) ![image](https://github.com/shadow/shadow/assets/96973603/4ea62e87-2d75-4af5-b2c6-1e8dab77caf7)

The left image depicts the results without the PoW-based defense mechanism, while the right image shows the results with the mechanism enabled. The format is as follows: (the number of connected clients / the number of all clients).

Could this issue be considered a bug? We're curious if Shadow can accurately simulate the PoW-based defense mechanism of Tor. Your insights on this matter would be greatly appreciated. Thank you for taking the time to read my question.

**Reply by jtracey** [reply]
PoW is the type of thing Shadow can't handle naively out of the box, because any computational PoW will be "instant" in simulation time. Shadow simulates I/O delays and, optionally, system call latency, not computation time -- see this [doc section on busy loops](https://shadow.github.io/docs/guide/limitations.html#busy-loops) (there should maybe be more direct documentation of how to roughly simulate computational overhead, unless I'm missing it somewhere). I haven't looked, but this smells like Tor trying to set a PoW hardness based on time, but the time to complete any hardness parameter will always be 0, so whenever the hardness gets above 0 for a connection, it'll just never converge. If that's true, the correct thing is probably a Tor patch that replaces the PoW algorithm with an equivalent duration `sleep()` call when run in testing mode.

That's just a theory for why it's manifesting the way it is, but regardless, you'll need to be careful how you test experiments that use PoW. This does seem like a bug, but it's not immediately obvious (to me at least) whether it would be a bug in Tor, Shadow, or, depending on whether this is one-off research or something you want to be widely usable, your setup.

**Reply by sporksmith** [reply]
> this smells like Tor trying to set a PoW hardness based on time,

I don't *think* the PoW hardness is set dynamically in this way... and I *think* we successfully used shadow when testing PoW. I'll see if I can look into this a bit...

**Reply by sporksmith** [reply]
Confirmed that we did some testing of PoW under shadow. It was a fairly minimal sanity check though, so wouldn't rule out the sort of "works, but with high error rates" behavior you're seeing.

If you haven't already, it might be worth trying the simulation with `--model-unblocked-syscall-latency=true`. This helps work around some of the sorts of issues that @jtracey was referring to.

**Reply by sporksmith** [reply]
The `stdout` of the `tor` process on the hidden service(s) would also be helpful, especially any warnings or errors.

One of the other tor devs suggested it'd be helpful to have some data from the metricsport, particularly `tor_hs_pow_suggested_effort`. I don't think tornettools currently supports this, but a quick and dirty way to do it manually is to add to the hidden service's tor config:

```
MetricsPort 127.0.0.1:9035
MetricsPortPolicy accept 127.0.0.1
```

and in the shadow config, on the same host as the hidden service, something like:

```
    - path: bash 
      args: 
      - -c 
      - | 
        set -euo pipefail 
        while true; do 
          echo 
          date 
          curl http://127.0.0.1:9035/metrics 
          sleep 60 
        done 
      start_time: 960 
      expected_final_state: running
```

You'll need a shadow built @ head for this sort of bash script to work well.

**Reply by Allgot** [reply]
>PoW is the type of thing Shadow can't handle naively out of the box, because any computational PoW will be "instant" in simulation time. Shadow simulates I/O delays and, optionally, system call latency, not computation time -- see this [doc section on busy loops](https://shadow.github.io/docs/guide/limitations.html#busy-loops) (there should maybe be more direct documentation of how to roughly simulate computational overhead, unless I'm missing it somewhere). 

> If you haven't already, it might be worth trying the simulation with `--model-unblocked-syscall-latency=true`. This helps work around some of the sorts of issues that jtracey was referring to.

Thank you for sharing your insights! I'll give your suggestions a try and get back with the results including the stdout of one successful client, one failed client, and one hidden service. I'll also try using metricsport as suggested.

> I haven't looked, but this smells like Tor trying to set a PoW hardness based on time, but the time to complete any hardness parameter will always be 0, so whenever the hardness gets above 0 for a connection, it'll just never converge. If that's true, the correct thing is probably a Tor patch that replaces the PoW algorithm with an equivalent duration `sleep()` call when run in testing mode.

Regarding the speculation about Tor setting PoW hardness (*'effort'* in Tor's terminology) based on time, it appears this isn't the case as clarified by @sporksmith. Thank you for offering your theory to consider nonetheless.

(FYI: We've already conducted mathematical modeling of Tor's PoW scheme and substituted PoW calculations with a stochastic alternative using `sleep()` in a separate branch, for later use.)

**Reply by Allgot** [reply]
Regrettably, it appears that the issue has not been resolved.

We have attached two images as before, which show the fluctuation in the number of connected clients to the hidden service over time.
![image](https://github.com/shadow/shadow/assets/96973603/746df959-be26-4457-8789-32a025cd0a3c) ![image](https://github.com/shadow/shadow/assets/96973603/6f9e56dd-f487-4af2-b26d-32b311d6575d)

We have also attached two log files (stdout) for each experiment. Unfortunately, we were unable to retrieve the data using metricsport due to a deadlock issue. However, you can still deduce the suggested effort from the stdout files.
[logs.zip](https://github.com/shadow/shadow/files/14468075/logs.zip)

Below are the versions of the tools utilized:
- shadow: commit (main) `72bd35f81b551b6f54d278d8e97a6c55648d371e`
- tor: commit (main) `0cccc7223c52512b5e0f47819f1a02740e5dad9b`
- tor (modified, used only for hidden services): The suggested effort is initialized as 300; all the other code remain consistent with commit (main) `0cccc7223c52512b5e0f47819f1a02740e5dad9b`

**Reply by stevenengler** [reply]
What do the numbers mean in your images? For example in `a: b/c`, what do each a, b, and c represent? What does "simultaneous access" mean?

> Specifically, while clients can normally establish and maintain connections to the hidden service when this mechanism is inactive, they fail to maintain their connection after the successful initial connection when the mechanism is activated.

What do you mean by "maintain their connection", do the SOCKS clients establish a working connection to the hidden service, but then that existing connection suddenly stops working? And what is meant by "initial connection"? How do these terms relate to the images above? I'm a bit confused about what exactly is happening.

If you can make a minimal example (with just a few relays and clients, small enough to run quickly on a laptop) that demonstrates the problem, it might help.

**Reply by Allgot** [reply]
> The left image depicts the results without the PoW-based defense mechanism, while the right image shows the results with the mechanism enabled. The format is as follows: (the number of connected clients / the number of all clients).

Apologies for the oversight. Let me provide further elaboration on the result with the above caption. We analyzed tgen's log and extracted information related to the successes and failures experienced by clients. For instance, if Client 1 initially succeeded but subsequently encountered failure, the result would be as below. 
```
0: 1/1
1: 0/1
```
The corresponding tgen log would look like the following.
```
...
... [message] [tgen-stream.c:1738] [_tgenstream_log] [stream-success] transport ...
... [message] [tgen-stream.c:1721] [_tgenstream_log] [stream-error] transport ...
...
```
I hope that this explanation would aid in clarifying the situation. When we mention "clients cannot maintain their connection", it means that they initially succeeded in transporting the stream (`stream-success`) but encountered an abrupt failure (`stream-error`) subsequently.

---
> If you can make a minimal example (with just a few relays and clients, small enough to run quickly on a laptop) that demonstrates the problem, it might help.

Because we utilized [torenttools](https://github.com/shadow/tornettools), I believe that replicating the issue using this tool should be straightforward. Below, I present the complete set of commands for staging, generating, and simulating.
```
# Staging
tornettools stage consensuses-2023-04 server-descriptors-2023-04 userstats-relay-country.csv tmodel-ccs2018.github.io --onionperf_data_path onionperf-2023-04 --bandwidth_data_path bandwidth-2023-04.csv --geoip_path tor/src/config/geoip

# Generation
tornettools generate relayinfo_staging_2023-04-01--2023-04-30.json userinfo_staging_2023-04-01--2023-04-30.json networkinfo_staging.gml tmodel-ccs2018.github.io/ --network_scale 0.01 --process_scale 0.01 --torperf_num_exit 0 --torperf_num_onion_service 0 --onion_service_user_scale 0.5 --exit_user_scale 0.5 --prefix tornet-0.01

# Simulation
tornettools simulate -a "--model-unblocked-syscall-latency=false --parallelism=48 --seed=666 --template-directory=shadow.data.template" tornet-0.01
(or)
tornettools simulate -a "--model-unblocked-syscall-latency=true --parallelism=48 --seed=666 --template-directory=shadow.data.template" tornet-0.01
```
If you find that this information is insufficient, please feel free to request any additional details. As I am relatively new to this field, errors may occur, and some important information could be missing. Thank you for your understanding.

**Reply by sporksmith** [reply]
Thanks for the follow-up info! I poked around a little through the logs but didn't immediately spot any obvious problems.

It looks like the suggested  effort gradually decreases from the initial 300.

```
$ grep update_suggested_effort hs.tor.1000.stdout\(pow-on\)
Jan 01 00:09:13.156 [info] update_suggested_effort(): Recalculated suggested effort: 200
Jan 01 00:14:13.311 [info] update_suggested_effort(): Recalculated suggested effort: 133
Jan 01 00:19:13.456 [info] update_suggested_effort(): Recalculated suggested effort: 88
Jan 01 00:24:13.612 [info] update_suggested_effort(): Recalculated suggested effort: 58
Jan 01 00:29:13.761 [info] update_suggested_effort(): Recalculated suggested effort: 38
Jan 01 00:34:13.916 [info] update_suggested_effort(): Recalculated suggested effort: 25
Jan 01 00:39:13.069 [info] update_suggested_effort(): Recalculated suggested effort: 16
Jan 01 00:44:13.219 [info] update_suggested_effort(): Recalculated suggested effort: 10
Jan 01 00:49:13.372 [info] update_suggested_effort(): Recalculated suggested effort: 6
Jan 01 00:54:13.520 [info] update_suggested_effort(): Recalculated suggested effort: 4
Jan 01 00:59:13.675 [info] update_suggested_effort(): Recalculated suggested effort: 2
```

The clients seem to be correctly using the suggested effort in their requests:

```
$ grep enqueue_rend_request hs.tor.1000.stdout\(pow-on\)
Jan 01 00:05:10.989 [info] enqueue_rend_request(): Enqueued rendezvous request with effort: 300. Queued requests: 1
Jan 01 00:05:14.559 [info] enqueue_rend_request(): Enqueued rendezvous request with effort: 300. Queued requests: 1
...
Jan 01 00:09:16.804 [info] enqueue_rend_request(): Enqueued rendezvous request with effort: 200. Queued requests: 1
Jan 01 00:09:44.144 [info] enqueue_rend_request(): Enqueued rendezvous request with effort: 200. Queued requests: 1
...
```

I'll continue looking into this...

**Reply by Allgot** [reply]
Oh, I think sharing the 'tgen' logs alongside the Tor logs could really help you figure out what's going on. I have attached the log files again. Thanks so much for your attention and effort!

[logs2.zip](https://github.com/shadow/shadow/files/14505372/logs2.zip)

**Reply by sporksmith** [reply]
I'm seeing a lot of error responses coming back from the intro points. e.g.

```
Jan 01 00:06:02.017 [info] handle_introduce_ack_bad(): Received INTRODUCE_ACK nack by $8A269A69067A353059B3C24C0316A3DCA8B3CE19~ [VJtO38jK4XYxnRFX7LOhHTf+kJaGynhhbeeJ21rk30A] at 202.61.225.95. Reason: 1
```

Reason 1 is ["Failure: service ID not recognized"](https://spec.torproject.org/rend-spec/introduction-protocol.html?highlight=introduce1#INTRO_ACK).

Surprisingly, there are more of these in the pow-off logs:

```
$ grep -c handle_introduce_ack_bad client*.tor.*
client2.tor.1000.stdout(pow-off):951
client2.tor.1000.stdout(pow-on):128
client.tor.1000.stdout(pow-off):389
client.tor.1000.stdout(pow-on):90
```

I noticed that the intro-point dos defense is enabled in both the pow and non-pow configurations:

```
Jan 01 00:04:00.000 [info] config_service_v3(): Service INTRO2 DoS defenses rate set to: 25
Jan 01 00:04:00.000 [info] config_service_v3(): Service INTRO2 DoS defenses burst set to: 200
```

I'm suspicious that this will cause problems in a tornettools-generated simulation, because it uses a small number of servers and clients to simulate a larger number (see e.g. `--process_scale`).

It does look like the intro point returns status=1 in the case that the INTRO2 DoS defense kicks in (among other reasons): https://gitlab.torproject.org/tpo/core/tor/-/blob/main/src/feature/hs/hs_intropoint.c?ref_type=heads#L703

I'm not sure why this would effect the pow-on case more than the pow-off case, but if nothing else may be something else going wrong that's confusing things.

@Allgot , could you see what's getting logged from `handle_introduce1` in the (intro point) relays? Something like `rg handle_introduce1 shadow.data` should do it. Could you also give your tor patch, and any other customization you've done? (e.g. did you also modify the tor config files to enable the DoS defenses?) I'll try to reproduce locally.

**Reply by sporksmith** [reply]
Actually the INTRO2 DoS defense may be a red herring. I've partially reproduced using the same `generate`  options, and overriding the default tor log level to `info` and setting `HiddenServiceEnableIntroDoSDefense 1`. All the `handle_introduce1` errors I see look like:

```
5294:Jan 01 00:05:21.858 [info] handle_introduce1(): No intro circuit found for INTRODUCE1 cell with auth key df0+MAxHZZTPzG4LFC+Pdu1r3mPSRMl6d5GGttl1wmQ from circuit 1033772687. Responding with NACK.
```

No sign of the INTRO2 DoS defense being triggered so far.

Still digging (but more-detailed repro instructions would still be useful)...

**Reply by sporksmith** [reply]
Related: https://gitlab.torproject.org/tpo/core/tor/-/issues/40922

I'm not sure yet whether that problem is related to the higher error rates with PoW enabled, and I probably won't be able to investigate further until next week, but disabling the rapid intro-point expirations at least cuts down on some of the other failure noise.

```
diff --git a/src/feature/hs/hs_service.c b/src/feature/hs/hs_service.c
index 3cc8c23e0b..f8d07e04cb 100644
--- a/src/feature/hs/hs_service.c
+++ b/src/feature/hs/hs_service.c
@@ -410,11 +410,6 @@ get_intro_point_max_introduce2(void)
 static int32_t
 get_intro_point_min_lifetime(void)
 {
-#define MIN_INTRO_POINT_LIFETIME_TESTING 10
-  if (get_options()->TestingTorNetwork) {
-    return MIN_INTRO_POINT_LIFETIME_TESTING;
-  }
-
   /* The [0, 2147483647] range is quite large to accommodate anything we decide
    * in the future. */
   return networkstatus_get_param(NULL, "hs_intro_min_lifetime",
@@ -427,11 +422,6 @@ get_intro_point_min_lifetime(void)
 static int32_t
 get_intro_point_max_lifetime(void)
 {
-#define MAX_INTRO_POINT_LIFETIME_TESTING 30
-  if (get_options()->TestingTorNetwork) {
-    return MAX_INTRO_POINT_LIFETIME_TESTING;
-  }
-
   /* The [0, 2147483647] range is quite large to accommodate anything we decide
    * in the future. */
   return networkstatus_get_param(NULL, "hs_intro_max_lifetime",

```

**Reply by sporksmith** [reply]
With PoW enabled, and with the high intro point turn-over, I had quite a few `stream-error`s in the first 8 sim minutes. With the above patch to disable high intro point turn-over, I had none in the first 8 minutes.

Didn't have a chance yet to do this comparison without PoW enabled, so still not sure this particular issue is worse in the PoW case or if there's something else going on.

**Reply by Allgot** [reply]
Wow, I appreciate your thorough investigation. Let me outline the entire experimentation process so that you can easily and precisely replicate our conditions. (I have manually verified that this accurately reproduces the problem!)
```
# version information
# shadow: commit (main) 72bd35f81b551b6f54d278d8e97a6c55648d371e
# tor: commit (main) 0cccc7223c52512b5e0f47819f1a02740e5dad9b
# !!You don't need to modify tor!!

# Staging
tornettools stage consensuses-2023-04 server-descriptors-2023-04 userstats-relay-country.csv tmodel-ccs2018.github.io --onionperf_data_path onionperf-2023-04 --bandwidth_data_path bandwidth-2023-04.csv --geoip_path tor/src/config/geoip

# Generation
tornettools generate relayinfo_staging_2023-04-01--2023-04-30.json userinfo_staging_2023-04-01--2023-04-30.json networkinfo_staging.gml tmodel-ccs2018.github.io/ --network_scale 0.01 --process_scale 0.01 --torperf_num_exit 0 --torperf_num_onion_service 0 --onion_service_user_scale 0.5 --exit_user_scale 0.5 --prefix repro

# Simulation
tornettools simulate -a "--model-unblocked-syscall-latency=false --parallelism=48 --seed=666 --template-directory=shadow.data.template" repro

# Save results
mv repro/shadow.data repro/pow_off
mv repro/shadow.log repro/pow_off.log

# Turn PoW on
echo "HiddenServicePoWDefensesEnabled 1" >> repro/conf/tor.onionservice.torrc

# Simulation
tornettools simulate -a "--model-unblocked-syscall-latency=false --parallelism=48 --seed=666 --template-directory=shadow.data.template" repro

# Save results
mv repro/shadow.data repro/pow_on
mv repro/shadow.log repro/pow_on.log

```

----
Following your suggested patch, we conducted the same experiments as before. Attached are two images displaying the results, and it seems that the problem has been solved. Thanks! (However, I still don't know why there are always some failing clients.)

![image](https://github.com/shadow/shadow/assets/96973603/0dc98960-10fd-46d3-8fb3-88a8fc06b289) ![image](https://github.com/shadow/shadow/assets/96973603/bebc0dcc-ec0c-4e29-adab-66ba186bf832)


----
> I'm suspicious that this will cause problems in a tornettools-generated simulation, because it uses a small number of servers and clients to simulate a larger number (see e.g. `--process_scale`).

Then can we conclude that increasing the process_scale towards 1 is the correct approach for conducting accurate experiments?

Once more, thank you for your detailed investigation.

**Reply by sporksmith** [reply]
> Then can we conclude that increasing the process_scale towards 1 is the correct approach for conducting accurate experiments?

It depends exactly what you're trying to accomplish. If you just don't want the intro-point-dos-defense to trigger, you can just not enable it (the tornettools-generated configs shouldn't enable it by default).

If you are trying to answer questions about what behavior might be like on the real network in the face of some particular load or DoS, then you'll need to think carefully about how to model that. `tornettools generate --help` includes many options that may be relevant, including `--process_scale`, `--server_scale`, and `--load_scale`

**Reply by sporksmith** [reply]
> Related: https://gitlab.torproject.org/tpo/core/tor/-/issues/40922

this issue is now fixed at tor HEAD. https://gitlab.torproject.org/tpo/core/tor/-/merge_requests/802

**Reply by Allgot** [reply]
> > Related: https://gitlab.torproject.org/tpo/core/tor/-/issues/40922
> 
> this issue is now fixed at tor HEAD. https://gitlab.torproject.org/tpo/core/tor/-/merge_requests/802

I'm glad to hear that our report was helpful in addressing the bug! I believe we can now properly set up our experiments. Thank you :)
